In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [5]:
!pip install pandas openpyxl


   -------------------- ------------------- 1/2 [openpyxl]
   -------------------- ------------------- 1/2 [openpyxl]
   -------------------- ------------------- 1/2 [openpyxl]
   ---------------------------------------- 2/2 [openpyxl]



In [7]:
df = pd.read_excel("../data/Telco_customer_churn.xlsx")
df.head()

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,...,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,...,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,...,Month-to-month,Yes,Electronic check,99.65,820.5,Yes,1,86,5372,Moved
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,...,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,...,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.3,Yes,1,89,5340,Competitor had better devices


In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 33 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CustomerID         7043 non-null   str    
 1   Count              7043 non-null   int64  
 2   Country            7043 non-null   str    
 3   State              7043 non-null   str    
 4   City               7043 non-null   str    
 5   Zip Code           7043 non-null   int64  
 6   Lat Long           7043 non-null   str    
 7   Latitude           7043 non-null   float64
 8   Longitude          7043 non-null   float64
 9   Gender             7043 non-null   str    
 10  Senior Citizen     7043 non-null   str    
 11  Partner            7043 non-null   str    
 12  Dependents         7043 non-null   str    
 13  Tenure Months      7043 non-null   int64  
 14  Phone Service      7043 non-null   str    
 15  Multiple Lines     7043 non-null   str    
 16  Internet Service   7043 non-null   

In [9]:
df.isnull().sum()

CustomerID              0
Count                   0
Country                 0
State                   0
City                    0
Zip Code                0
Lat Long                0
Latitude                0
Longitude               0
Gender                  0
Senior Citizen          0
Partner                 0
Dependents              0
Tenure Months           0
Phone Service           0
Multiple Lines          0
Internet Service        0
Online Security         0
Online Backup           0
Device Protection       0
Tech Support            0
Streaming TV            0
Streaming Movies        0
Contract                0
Paperless Billing       0
Payment Method          0
Monthly Charges         0
Total Charges           0
Churn Label             0
Churn Value             0
Churn Score             0
CLTV                    0
Churn Reason         5174
dtype: int64

In [10]:
df.shape

(7043, 33)

In [11]:
df.columns.tolist()

['CustomerID',
 'Count',
 'Country',
 'State',
 'City',
 'Zip Code',
 'Lat Long',
 'Latitude',
 'Longitude',
 'Gender',
 'Senior Citizen',
 'Partner',
 'Dependents',
 'Tenure Months',
 'Phone Service',
 'Multiple Lines',
 'Internet Service',
 'Online Security',
 'Online Backup',
 'Device Protection',
 'Tech Support',
 'Streaming TV',
 'Streaming Movies',
 'Contract',
 'Paperless Billing',
 'Payment Method',
 'Monthly Charges',
 'Total Charges',
 'Churn Label',
 'Churn Value',
 'Churn Score',
 'CLTV',
 'Churn Reason']

In [12]:
df["Churn Value"].sum()

np.int64(1869)

In [16]:
df.groupby("Churn Label")['Churn Score'].describe()

,count,mean,std,min,25%,50%,75%,max
Churn Label,,,,,,,,
No,5174.0,50.098183,17.702277,5.0,35.0,50.0,66.0,80.0
Yes,1869.0,82.510433,10.328570,65.0,74.0,82.0,91.0,100.0


In [23]:
df.groupby('Churn Label')['Churn Value'].describe()

,count,mean,std,min,25%,50%,75%,max
Churn Label,,,,,,,,
No,5174.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
Yes,1869.0,1.0,0.0,1.0,1.0,1.0,1.0,1.0


In [25]:
df['Churn Value'].sum()
df.groupby('Churn Label')['Churn Score'].describe()

,count,mean,std,min,25%,50%,75%,max
Churn Label,,,,,,,,
No,5174.0,50.098183,17.702277,5.0,35.0,50.0,66.0,80.0
Yes,1869.0,82.510433,10.328570,65.0,74.0,82.0,91.0,100.0


checking what is unique or not

In [26]:
df[['State', 'City', 'Zip Code', 'Latitude', 'Longitude']].nunique()

State           1
City         1129
Zip Code     1652
Latitude     1652
Longitude    1651
dtype: int64

In [27]:
df['State'].nunique()

1

In [32]:
df['City'].nunique()
df['City'].value_counts().head(10)

City
Los Angeles      305
San Diego        150
San Jose         112
Sacramento       108
San Francisco    104
Fresno            64
Long Beach        60
Oakland           52
Stockton          44
Glendale          40
Name: count, dtype: int64

## Phase 1: Column Decisions

**Drop (identifiers / no signal):**
- CustomerID — unique identifier, no predictive value
- Count — constant column, no signal
- Country — single value across dataset, no signal
- State — single value across dataset, no signal

**Drop (redundant geography):**
- Lat Long, Latitude, Longitude, Zip Code — redundant with each other; 
  city-level signal will be captured via City (bucketed) instead

**Defer to EDA / preprocessing:**
- City — high cardinality (1,129 unique values); will bucket top-N + "Other" 
  in Phase 3 if EDA shows meaningful churn-rate variation by city

**Drop (target leakage):**
- Churn Label — duplicate of Churn Value (text vs numeric), redundant
- Churn Score — likely an existing model's output, wouldn't exist at prediction time
- Churn Reason — only populated for already-churned customers (1,869 non-null 
  matches churned count), couldn't exist before the outcome
  
**Keep (legitimate feature):**
- CLTV — modest distributional difference by churn (means 4491 vs 4149, 
  overlapping ranges), unlike Churn Score/Reason; represents a value 
  calculable for any active customer, not derived from the churn outcome

**Target:**
- Churn Value (0/1) — keep, drop Churn Label

In [33]:
df.groupby('Churn Value')['CLTV'].describe()

,count,mean,std,min,25%,50%,75%,max
Churn Value,,,,,,,,
0,5174.0,4490.921337,1167.703198,2003.0,3643.75,4620.0,5434.75,6500.0
1,1869.0,4149.414660,1189.370707,2003.0,3101.00,4238.0,5166.00,6484.0


## Phase 1: Final Column Decisions

Drop:
- CustomerID, Count       (identifier / constant, no signal)
- Country, State           (constant, no signal)
- Lat Long, Latitude, Longitude, Zip Code   (redundant geography)
- Churn Label               (duplicate of Churn Value)
- Churn Score, Churn Reason  (target leakage)

Keep as feature:
- CLTV   (correlated but not derived from outcome)
- City   (deferred — will bucket top-N + "Other" in Phase 3 if EDA supports it)
- All service/account/demographic columns (Gender, Partner, Tenure Months, 
  Contract, Internet Service, etc.)

Target:
- Churn Value (0/1)

In [34]:
cols_to_drop = ['CustomerID', 'Count', 'Country', 'State', 
                 'Lat Long', 'Latitude', 'Longitude', 'Zip Code',
                 'Churn Label', 'Churn Score', 'Churn Reason']

df_clean = df.drop(columns=cols_to_drop)
df_clean.shape
df_clean.columns.tolist()

['City',
 'Gender',
 'Senior Citizen',
 'Partner',
 'Dependents',
 'Tenure Months',
 'Phone Service',
 'Multiple Lines',
 'Internet Service',
 'Online Security',
 'Online Backup',
 'Device Protection',
 'Tech Support',
 'Streaming TV',
 'Streaming Movies',
 'Contract',
 'Paperless Billing',
 'Payment Method',
 'Monthly Charges',
 'Total Charges',
 'Churn Value',
 'CLTV']

In [35]:
df_clean.isnull().sum()

City                 0
Gender               0
Senior Citizen       0
Partner              0
Dependents           0
Tenure Months        0
Phone Service        0
Multiple Lines       0
Internet Service     0
Online Security      0
Online Backup        0
Device Protection    0
Tech Support         0
Streaming TV         0
Streaming Movies     0
Contract             0
Paperless Billing    0
Payment Method       0
Monthly Charges      0
Total Charges        0
Churn Value          0
CLTV                 0
dtype: int64

In [36]:
df_clean['Total Charges'].dtype

dtype('O')

In [37]:
pd.to_numeric(df_clean['Total Charges'], errors='coerce').isnull().sum()

np.int64(11)

In [38]:
mask = pd.to_numeric(df_clean['Total Charges'], errors='coerce').isnull()
df_clean[mask][['Tenure Months', 'Monthly Charges', 'Total Charges']]

,Tenure Months,Monthly Charges,Total Charges
2234,0,52.55,
2438,0,20.25,
2568,0,80.85,
2667,0,25.75,
2856,0,56.05,
4331,0,19.85,
4687,0,25.35,
5104,0,20.00,
5719,0,19.70,
6772,0,73.35,


In [39]:
df_clean['Total Charges'] = pd.to_numeric(df_clean['Total Charges'], errors='coerce')
df_clean['Total Charges'] = df_clean['Total Charges'].fillna(0)
df_clean['Total Charges'].isnull().sum()   # should be 0
df_clean['Total Charges'].dtype             # should be float64

dtype('float64')

In [ ]:
df_clean.to_excel("../data/telco_churn_clean.xlsx", index=False)
